File: ruby/dialog/serverdialog.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'dialoghelper'

class ServerDialog < OSX::NSObject
  include OSX
  include DialogHelper  
  attr_accessor :delegate, :prefix, :parent
  attr_reader :uid
  ib_outlet :window
  ib_mapped_outlet :nameText, :hostCombo, :passwordText, :nickText, :usernameText, :realnameText, :auto_connectCheck
  ib_mapped_int_outlet :portText, :encodingCombo
  ib_mapped_outlet :leaving_commentText, :userinfoText, :invisibleCheck
  ib_mapped_outlet :login_commandsText
  ib_outlet :channelsTable, :addButton, :editButton, :deleteButton
  ib_outlet :okButton
  
  TABLE_ROW_TYPE = 'row'
  TABLE_ROW_TYPES = [TABLE_ROW_TYPE]
  
  def initialize
    @prefix = 'serverDialog'
  end
  
  def config
    @c
  end
  
  def start(conf, uid)
    @c = conf
    @uid = uid
    NSBundle.loadNibNamed_owner('ServerDialog', self)
    @channelsTable.setTarget(self)
    @channelsTable.setDoubleAction('tableView_doubleClicked:')
  	@channelsTable.registerForDraggedTypes(TABLE_ROW_TYPES);
    @window.setTitle("New Server") if uid < 0
    load
    update_connection_page
    update_channels_page
    show
  end
  
  def show
    @window.centerOfWindow(@parent) unless @window.isVisible
    @window.makeKeyAndOrderFront(self)
  end
  
  def close
    @channelsTable.unregisterDraggedTypes
    @delegate = nil
    @window.close
  end
  
  def windowWillClose(sender)
    fire_event('onClose')
  end
  
  def onOk(sender)
    save
    fire_event('onOk', @c)
    @window.close
  end
  
  def onCancel(sender)
    @window.close
  end
  
  def load
    load_mapped_outlets(@c)
  end
  
  def save
    save_mapped_outlets(@c)
    @c.login_commands.delete_if {|i| i =~ /^\s*$/ }
  end
  
  def controlTextDidChange(n)
    update_connection_page
  end
  
  def update_connection_page
    name = @nameText.stringValue.to_s
    host = @hostCombo.stringValue.to_s
    port = @portText.stringValue.to_s
    nick = @nickText.stringValue.to_s
    username = @usernameText.stringValue.to_s
    realname = @realnameText.stringValue.to_s
    @okButton.setEnabled(!name.empty? && !host.empty? && port.to_i > 0 && !nick.empty? && !username.empty? && !realname.empty?)
  end
  
  def update_channels_page
    t = @channelsTable
    sel = t.selectedRows[0]
    unless sel
      @editButton.setEnabled(false)
      @deleteButton.setEnabled(false)
    else
      @editButton.setEnabled(true)
      @deleteButton.setEnabled(true)
    end
  end
  
  def reload_table
    @channelsTable.reloadData
  end
  
  def numberOfRowsInTableView(sender)
    @c.channels.size
  end
  
  def tableView_objectValueForTableColumn_row(sender, col, row)
    i = @c.channels[row]
    col = col.identifier.to_s.to_sym
    case col
    when :name; i.name
    when :pass; i.password
    when :join; i.auto_join
    end
  end
  
  def tableView_setObjectValue_forTableColumn_row(sender, obj, col, row)
    i = @c.channels[row]
    col = col.identifier.to_s.to_sym
    case col
    when :join; i.auto_join = obj.intValue != 0
    end
  end
  
  def tableViewSelectionDidChange(n)
    update_channels_page
  end
  
  def tableView_doubleClicked(sender)
    onEdit(sender)
  end
  
  #objc_method :tableView_writeRows_toPasteboard, 'c@:@@@'
  def tableView_writeRows_toPasteboard(sender, rows, pboard)
    pboard.declareTypes_owner(TABLE_ROW_TYPES, self)
    pboard.setPropertyList_forType(rows, TABLE_ROW_TYPE)
    true
  end
  
  def tableView_validateDrop_proposedRow_proposedDropOperation(sender, info, row, op)
  	pboard = info.draggingPasteboard
  	if op == NSTableViewDropAbove && pboard.availableTypeFromArray(TABLE_ROW_TYPES)
  	  NSDragOperationGeneric
	  else
	    NSDragOperationNone
    end
  end
  
  #objc_method :tableView_acceptDrop_row_dropOperation, 'c@:@@ii'
  def tableView_acceptDrop_row_dropOperation(sender, info, row, op)
  	pboard = info.draggingPasteboard
  	return false unless op == NSTableViewDropAbove && pboard.availableTypeFromArray(TABLE_ROW_TYPES)
    ary = @c.channels
    sel = @channelsTable.selectedRows.map {|i| ary[i.to_i] }
    
    targets = pboard.propertyListForType(TABLE_ROW_TYPE).to_a.map {|i| ary[i.to_i] }
    low = ary[0...row] || []
    high = ary[row...ary.size] || []
    targets.each do |i|
      low.delete(i)
      high.delete(i)
    end
    @c.channels = low + targets + high

    unless sel.empty?
      sel = sel.map {|i| @c.channels.index(i) }
      @channelsTable.selectRows(sel)
    end
    reload_table
    true
  end

  def onAdd(sender)
    sel = @channelsTable.selectedRows[0]
    conf = sel ? @c.channels[sel] : IRCChannelConfig.new
    @sheet = ChannelDialog.alloc.init
    @sheet.delegate = self
    @sheet.start_sheet(@window, conf, true)
  end
  
  def onEdit(sender)
    sel = @channelsTable.selectedRows[0]
    return unless sel
    conf = @c.channels[sel]
    @sheet = ChannelDialog.alloc.init
    @sheet.delegate = self
    @sheet.start_sheet(@window, conf)
  end
  
  def channelDialog_onOk(sender, conf)
    i = @c.channels.index {|t| t.name == conf.name }
    if i
      @c.channels[i] = conf
    else
      @c.channels << conf
    end
    reload_table
    @sheet = nil
  end
  
  def channelDialog_onCancel(sender)
    @sheet = nil
  end
  
  def onDelete(sender)
    sel = @channelsTable.selectedRows[0]
    return unless sel
    @c.channels.delete_at(sel)
    count = @c.channels.size
    if count > 0
      if count <= sel
        @channelsTable.select(count - 1)
      else
        @channelsTable.select(sel)
      end
    end
    reload_table
  end
end

File: objc/Dialogs/ServerDialog.m
// Created by Satoshi Nakagawa.
// You can redistribute it and/or modify it under the Ruby's license or the GPL2.

#import "ServerDialog.h"
#import "NSWindowHelper.h"
#import "NSLocaleHelper.h"
#import "IRCChannelConfig.h"


#define TABLE_ROW_TYPE		@"row"
#define TABLE_ROW_TYPES		[NSArray arrayWithObject:TABLE_ROW_TYPE]


@interface ServerDialog (Private)
- (void)load;
- (void)save;
- (void)updateConnectionPage;
- (void)updateChannelsPage;
- (void)reloadChannelTable;
@end


@implementation ServerDialog

@synthesize delegate;
@synthesize parentWindow;
@synthesize uid;
@synthesize config;

- (id)init
{
	if (self = [super init]) {
		[NSBundle loadNibNamed:@"ServerDialog" owner:self];
		
		if ([NSLocale prefersJapaneseLanguage]) {
			[hostCombo addItemWithObjectValue:@"irc.ircnet.ne.jp (IRCnet)"];
			[hostCombo addItemWithObjectValue:@"-"];
			[hostCombo addItemWithObjectValue:@"irc.friend-chat.jp (Friend)"];
			[hostCombo addItemWithObjectValue:@"irc.2ch.net (2ch)"];
			[hostCombo addItemWithObjectValue:@"irc.cre.ne.jp (cre)"];
			[hostCombo addItemWithObjectValue:@"-"];
			[hostCombo addItemWithObjectValue:@"chat.freenode.net (freenode)"];
			[hostCombo addItemWithObjectValue:@"eu.undernet.org (Undernet)"];
			[hostCombo addItemWithObjectValue:@"irc.quakenet.org (QuakeNet)"];
			[hostCombo addItemWithObjectValue:@"chat1.ustream.tv (Ustream)"];
		}
		else {
			[hostCombo addItemWithObjectValue:@"chat.freenode.net (freenode)"];
			[hostCombo addItemWithObjectValue:@"irc.efnet.net (EFnet)"];
			[hostCombo addItemWithObjectValue:@"irc.us.ircnet.net (IRCnet)"];
			[hostCombo addItemWithObjectValue:@"irc.fr.ircnet.net (IRCnet)"];
			[hostCombo addItemWithObjectValue:@"us.undernet.org (Undernet)"];
			[hostCombo addItemWithObjectValue:@"eu.undernet.org (Undernet)"];
			[hostCombo addItemWithObjectValue:@"irc.quakenet.org (QuakeNet)"];
			[hostCombo addItemWithObjectValue:@"uk.quakenet.org (QuakeNet)"];
			[hostCombo addItemWithObjectValue:@"irc.mozilla.org (Mozilla)"];
			[hostCombo addItemWithObjectValue:@"chat1.ustream.tv (Ustream)"];
		}
	}
	return self;
}

- (void)dealloc
{
	[config release];
	[super dealloc];
}

- (void)start
{
	if (uid < 0) {
		[self.window setTitle:@"New Server"];
	}
	
	[channelTable setTarget:self];
	[channelTable setDoubleAction:@selector(tableViewDoubleClicked:)];
	[channelTable registerForDraggedTypes:TABLE_ROW_TYPES];

	[self load];
	[self updateConnectionPage];
	[self updateChannelsPage];
	[self encodingChanged:nil];
	[self proxyChanged:nil];
	[self reloadChannelTable];
	
	[self show];
}

- (void)show
{
	if (![self.window isVisible]) {
		[self.window centerOfWindow:parentWindow];
	}
	[self.window makeKeyAndOrderFront:nil];
}

- (void)load
{
	[nameText setStringValue:config.name];
	[autoConnectCheck setState:config.autoConnect];
	
	[hostCombo setStringValue:config.host];
	[sslCheck setState:config.useSSL];
	[portText setIntValue:config.port];
	
	[nickText setStringValue:config.nick];
	[passwordText setStringValue:config.password];
	[usernameText setStringValue:config.username];
	[realNameText setStringValue:config.realName];
	[nickPasswordText setStringValue:config.nickPassword];
	if (config.altNicks.count) {
		[altNicksText setStringValue:[config.altNicks componentsJoinedByString:@" "]];
	}
	else {
		[altNicksText setStringValue:@""];
	}

	[leavingCommentText setStringValue:config.leavingComment];
	[userInfoText setStringValue:config.userInfo];

	[encodingCombo selectItemWithTag:config.encoding];
	[fallbackEncodingCombo selectItemWithTag:config.fallbackEncoding];
	
	[proxyCombo selectItemWithTag:config.proxyType];
	[proxyHostText setStringValue:config.proxyHost];
	[proxyPortText setIntValue:config.proxyPort];
	[proxyUserText setStringValue:config.proxyUser];
	[proxyPasswordText setStringValue:config.proxyPassword];
	
	//IBOutlet ListView* channelsTable;

	[loginCommandsText setString:[config.loginCommands componentsJoinedByString:@"\n"]];
	[invisibleCheck setState:config.invisibleMode];
}

- (void)save
{
}

- (void)updateConnectionPage
{
	NSString* name = [nameText stringValue];
	NSString* host = [hostCombo stringValue];
	int port = [portText intValue];
	NSString* nick = [nickText stringValue];
	
	BOOL enabled = name.length > 0 && host.length > 0 && ![host isEqualToString:@"-"] && port > 0 && nick.length > 0;
	[okButton setEnabled:enabled];
}

- (void)updateChannelsPage
{
	int i = [channelTable selectedRow];
	BOOL enabled = (i >= 0);
	[editChannelButton setEnabled:enabled];
	[deleteChannelButton setEnabled:enabled];
}

- (void)reloadChannelTable
{
	[channelTable reloadData];
}

#pragma mark -
#pragma mark Actions

- (void)ok:(id)sender
{
	[self save];
	
	if ([delegate respondsToSelector:@selector(serverDialogOnOK:)]) {
		[delegate serverDialogOnOK:self];
	}
	
	[self.window close];
}

- (void)cancel:(id)sender
{
	[self.window close];
}

- (void)controlTextDidChange:(NSNotification*)note
{
	[self updateConnectionPage];
}

- (void)hostComboChanged:(id)sender
{
	[self updateConnectionPage];
}

- (void)encodingChanged:(id)sender
{
	int tag = [[encodingCombo selectedItem] tag];
	[fallbackEncodingCombo setEnabled:(tag == NSUTF8StringEncoding)];
}

- (void)proxyChanged:(id)sender
{
	int tag = [[proxyCombo selectedItem] tag];
	BOOL enabled = (tag == PROXY_SOCKS4 || tag == PROXY_SOCKS5);
	[proxyHostText setEnabled:enabled];
	[proxyPortText setEnabled:enabled];
	[proxyUserText setEnabled:enabled];
	[proxyPasswordText setEnabled:enabled];
	[sslCheck setEnabled:tag == PROXY_NONE];
}

- (void)addChannel:(id)sender
{
	LOG_METHOD
}

- (void)deleteChannel:(id)sender
{
	LOG_METHOD
}

- (void)editChannel:(id)sender
{
	LOG_METHOD
}

#pragma mark -
#pragma mark NSTableViwe Delegate

- (NSInteger)numberOfRowsInTableView:(NSTableView *)sender
{
	return config.channels.count;
}

- (id)tableView:(NSTableView *)sender objectValueForTableColumn:(NSTableColumn *)column row:(NSInteger)row
{
	IRCChannelConfig* c = [config.channels objectAtIndex:row];
	NSString* columnId = [column identifier];
	
	if ([columnId isEqualToString:@"name"]) {
		return c.name;
	}
	else if ([columnId isEqualToString:@"pass"]) {
		return c.password;
	}
	else if ([columnId isEqualToString:@"join"]) {
		return [NSNumber numberWithBool:c.autoJoin];
	}
	
	return nil;
}

- (void)tableView:(NSTableView *)sender setObjectValue:(id)obj forTableColumn:(NSTableColumn *)column row:(NSInteger)row
{
	IRCChannelConfig* c = [config.channels objectAtIndex:row];
	NSString* columnId = [column identifier];
	
	if ([columnId isEqualToString:@"join"]) {
		c.autoJoin = [obj intValue] != 0;
	}
}

- (void)tableViewSelectionDidChange:(NSNotification *)note
{
	[self updateChannelsPage];
}

- (void)tableViewDoubleClicked:(id)sender
{
	[self editChannel:nil];
}

- (BOOL)tableView:(NSTableView *)sender writeRowsWithIndexes:(NSIndexSet *)rows toPasteboard:(NSPasteboard *)pboard
{
	NSArray* ary = [NSArray arrayWithObject:[NSNumber numberWithInt:[rows firstIndex]]];
	
	[pboard declareTypes:TABLE_ROW_TYPES owner:self];
	[pboard setPropertyList:ary forType:TABLE_ROW_TYPE];
	return YES;
}

- (NSDragOperation)tableView:(NSTableView *)sender validateDrop:(id < NSDraggingInfo >)info proposedRow:(NSInteger)row proposedDropOperation:(NSTableViewDropOperation)op
{
	NSPasteboard* pboard = [info draggingPasteboard];
	if (op == NSTableViewDropAbove && [pboard availableTypeFromArray:TABLE_ROW_TYPES]) {
		return NSDragOperationGeneric;
	}
	else {
		return NSDragOperationNone;
	}
}

- (BOOL)tableView:(NSTableView *)sender acceptDrop:(id < NSDraggingInfo >)info row:(NSInteger)row dropOperation:(NSTableViewDropOperation)op
{
	NSPasteboard* pboard = [info draggingPasteboard];
	if (op == NSTableViewDropAbove && [pboard availableTypeFromArray:TABLE_ROW_TYPES]) {
		NSArray* selectedRows = [pboard propertyListForType:TABLE_ROW_TYPE];
		int sel = [[selectedRows objectAtIndex:0] intValue];
		
		NSMutableArray* ary = config.channels;
		IRCChannelConfig* target = [ary objectAtIndex:sel];
		[[target retain] autorelease];

		NSMutableArray* low = [[[ary subarrayWithRange:NSMakeRange(0, row)] mutableCopy] autorelease];
		NSMutableArray* high = [[[ary subarrayWithRange:NSMakeRange(row, ary.count - row)] mutableCopy] autorelease];
		
		[low removeObjectIdenticalTo:target];
		[high removeObjectIdenticalTo:target];
		
		[ary removeAllObjects];
		
		[ary addObjectsFromArray:low];
		[ary addObject:target];
		[ary addObjectsFromArray:high];
		
		[self reloadChannelTable];
		
		sel = [ary indexOfObjectIdenticalTo:target];
		if (0 <= sel) {
			[channelTable select:sel];
		}
		
		return YES;
	}
	return NO;
}

#pragma mark -
#pragma mark NSWindow Delegate

- (void)windowWillClose:(NSNotification*)note
{
	[channelTable unregisterDraggedTypes];
	
	if ([delegate respondsToSelector:@selector(serverDialogWillClose:)]) {
		[delegate serverDialogWillClose:self];
	}
}

@end
